jlm/llvm/ir/operators/StoreVolatile.hpp

```diff
@@ -0,0 +1,219 @@
+ */
+class StoreVolatileOperation final : public rvsdg::simple_op
+{
+public:
+  ~StoreVolatileOperation() noexcept override;
+
+  StoreVolatileOperation(
+      const rvsdg::valuetype & storedType,
+      size_t numMemoryStates,
+      size_t alignment)
+      : simple_op(
+          CreateOperandPorts(storedType, numMemoryStates),
+          CreateResultPorts(numMemoryStates)),
+        Alignment_(alignment)
+  {}
+
+  bool
+  operator==(const operation & other) const noexcept override;
+
+  [[nodiscard]] std::string
+  debug_string() const override;
+
+  [[nodiscard]] std::unique_ptr<rvsdg::operation>
+  copy() const override;
+
+  [[nodiscard]] const jlm::rvsdg::valuetype &
+  GetStoredType() const noexcept
+  {
+    return *util::AssertedCast<const jlm::rvsdg::valuetype>(&argument(1).type());
+  }
+
+  [[nodiscard]] size_t
+  NumMemoryStates() const noexcept
+  {
+    return nresults() - 1;
+  }
+
+  [[nodiscard]] size_t
+  GetAlignment() const noexcept
+  {
+    return Alignment_;
+  }
+
+  static std::unique_ptr<llvm::tac>
+  Create(
+      const variable * address,
+      const variable * value,
+      const variable * ioState,
+      const variable * memoryState,
+      size_t alignment)
+  {
+    auto & storedType = CheckAndExtractStoredType(value->type());
+
+    StoreVolatileOperation op(storedType, 1, alignment);
+    return tac::create(op, { address, value, ioState, memoryState });
+  }
+
+private:
+  static const rvsdg::valuetype &
+  CheckAndExtractStoredType(const rvsdg::type & type)
+  {
+    if (auto storedType = dynamic_cast<const rvsdg::valuetype *>(&type))
+      return *storedType;
+
+    throw jlm::util::error("Expected value type");
+  }
+
+  static std::vector<rvsdg::port>
+  CreateOperandPorts(const rvsdg::valuetype & storedType, size_t numStates)
+  {
+    std::vector<rvsdg::port> ports({ PointerType(), storedType, iostatetype() });
+    std::vector<rvsdg::port> states(numStates, { MemoryStateType() });
+    ports.insert(ports.end(), states.begin(), states.end());
+    return ports;
+  }
+
+  static std::vector<rvsdg::port>
+  CreateResultPorts(size_t numMemoryStates)
+  {
+    std::vector<rvsdg::port> ports({ iostatetype() });
+    std::vector<rvsdg::port> memoryStates(numMemoryStates, { MemoryStateType() });
+    ports.insert(ports.end(), memoryStates.begin(), memoryStates.end());
+    return ports;
+  }
+
+  size_t Alignment_;
+};
+
+/**
+ * Represents a StoreVolatileOperation in an RVSDG.
+ */
+class StoreVolatileNode final : public rvsdg::simple_node
+{
+  StoreVolatileNode(
+      rvsdg::region & region,
+      const StoreVolatileOperation & operation,
+      const std::vector<rvsdg::output *> & operands)
+      : simple_node(&region, operation, operands)
+  {}
+
+public:
+  [[nodiscard]] const StoreVolatileOperation &
+  GetOperation() const noexcept
+  {
+    return *util::AssertedCast<const StoreVolatileOperation>(&operation());
+  }
+
+  [[nodiscard]] size_t
+  NumMemoryStates() const noexcept
```

Comment:
Are there no helper functions for `GetMemoryStateInput(index)` and `GetMemoryStateOutput(index)`?

Reply:
Meh, I forgot them. It was also not needed when writing the tests. I will add them once they are required. I want to get this merged and move on.